File: parent/formula-translator/src/main/java/antlr/test/cal/CalMain.java
package antlr.test.cal;

import org.antlr.v4.runtime.*;
import org.antlr.v4.runtime.tree.ErrorNode;
import org.antlr.v4.runtime.tree.ParseTree;
import org.antlr.v4.runtime.tree.ParseTreeWalker;
import org.antlr.v4.runtime.tree.TerminalNode;

import java.util.HashMap;
import java.util.Map;

public class CalMain {
    public static void main(String[] args) throws Exception {

        String input ="193\n" + "a=5\n" + "b=6\n"+"a+b*2\n"+"(1+2)*3\n";
        CharStream charStream = CharStreams.fromString(input);
        CalLexer lexer = new CalLexer(charStream);
        TokenStream tokenStream = new CommonTokenStream(lexer);
        CalParser parser = new CalParser(tokenStream);
        ParseTree tree = parser.prog();
//        ParseTreeWalker walker = new ParseTreeWalker();
//
//        walker.walk(new MyListener(), tree);
        EvalVisitor visitor = new EvalVisitor();
        visitor.visit(tree);
    }

    static class MyListener extends CalBaseListener {

    }

    static class EvalVisitor extends CalBaseVisitor<Integer> {
        Map<String, Integer> memory = new HashMap<>();

        @Override
        public Integer visitPrintExpr(CalParser.PrintExprContext ctx) {
            int value = this.visit(ctx.expr());
            System.out.println(value);
            return 0;
        }

        /**
         * ID '=' expr NEWLINE
         */
        @Override
        public Integer visitAssign(CalParser.AssignContext ctx) {
            String id = ctx.ID().getText();
            int value = this.visit(ctx.expr());
            memory.put(id,value);
            return value;
        }

        @Override
        public Integer visitBlank(CalParser.BlankContext ctx) {
            return super.visitBlank(ctx);
        }

        @Override
        public Integer visitParens(CalParser.ParensContext ctx) {
            return this.visit(ctx.expr());
        }

        @Override
        public Integer visitMulDiv(CalParser.MulDivContext ctx) {
            int left = visit(ctx.expr(0));
            int right = this.visit(ctx.expr(1));
            if(ctx.op.getType() == CalParser.MUL){
                return left * right;
            }else{
                return left/right;
            }
        }

        @Override
        public Integer visitAddSub(CalParser.AddSubContext ctx) {
            int left = this.visit(ctx.expr(0));
            int right = this.visit(ctx.expr(1));
            if(ctx.op.getType() == CalParser.ADD){
                return left + right;
            }else{
                return left - right;
            }
        }

        // ID
        @Override
        public Integer visitId(CalParser.IdContext ctx) {
            String id = ctx.ID().getText();
            if(memory.containsKey(id)){
                return memory.get(id);
            }else{
                return 0;
            }
        }

        // INT
        @Override
        public Integer visitInt(CalParser.IntContext ctx) {
            return Integer.valueOf(ctx.INT().getText());
        }
    }

}

File: parent/formula-translator/src/main/kotlin/com/github/xadkile/p6/formula/translator/antlr/eg/CalMain.kt
package com.github.xadkile.p6.formula.translator.antlr.eg

import com.github.xadkile.p6.formula.translator.antlr.eg.CalParser.*
import org.antlr.v4.runtime.CharStream
import org.antlr.v4.runtime.CharStreams
import org.antlr.v4.runtime.CommonTokenStream
import org.antlr.v4.runtime.TokenStream
import org.antlr.v4.runtime.tree.ParseTree

object CalMain {
    @Throws(Exception::class)
    @JvmStatic
    fun main(args: Array<String>) {
        val input = """
            193
            a=5
            b=6
            a+b*2
            (1+2)*3

            """.trimIndent()
        val charStream: CharStream = CharStreams.fromString(input)
        val lexer = CalLexer(charStream)
        val tokenStream: TokenStream = CommonTokenStream(lexer)
        val parser = CalParser(tokenStream)
        val tree: ParseTree = parser.prog()
        val visitor = EvalVisitor()
        visitor.visit(tree)
    }

    internal class MyListener : CalBaseListener() {
        override fun enterProg(ctx: CalParser.ProgContext?) {
            super.enterProg(ctx)
        }
    }

    internal class EvalVisitor : CalBaseVisitor<Int?>() {
        var memory: MutableMap<String, Int> = HashMap()
        override fun visitClear(ctx: CalParser.ClearContext?): Int {
            memory.clear()
            return 0
        }

        override fun visitPrintExpr(ctx: CalParser.PrintExprContext): Int {
            val value: Int? = this.visit(ctx.expr())
            println(value)
            return 0
        }

        /**
         * ID '=' expr NEWLINE
         */
        override fun visitAssign(ctx: CalParser.AssignContext): Int {
            val id = ctx.ID().text
            val value: Int? = this.visit(ctx.expr())
            memory[id] = value ?: 0
            return value ?: 0
        }

        override fun visitBlank(ctx: CalParser.BlankContext?): Int {
            return super.visitBlank(ctx) ?: 0
        }

        override fun visitParens(ctx: ParensContext): Int {
            return this.visit(ctx.expr()) ?:0
        }

        override fun visitMulDiv(ctx: MulDivContext): Int {
            val left: Int = visit(ctx.expr(0)) ?:0
            val right: Int = this.visit(ctx.expr(1)) ?:0
            return if (ctx.op.type == MUL) {
                left * right
            } else {
                left / right
            }
        }

        override fun visitAddSub(ctx: AddSubContext): Int {
            val left: Int = this.visit(ctx.expr(0)) ?:0
            val right: Int = this.visit(ctx.expr(1))?:0
            return if (ctx.op.type == ADD) {
                left + right
            } else {
                left - right
            }
        }

        // ID
        override fun visitId(ctx: IdContext): Int? {
            val id = ctx.ID().text
            return if (memory.containsKey(id)) {
                memory[id]
            } else {
                0
            }
        }

        // INT
        override fun visitInt(ctx: IntContext): Int {
            return Integer.valueOf(ctx.INT().text)
        }
    }
}
